modules/saku-policy-store/src/saku/schemas.clj

```diff
@@ -1,53 +1,198 @@
+                       [k-src rename])))
+             (m/entries schema))))}}}))
 
-(def identity-policies
-  (m/schema [:sequential identity-policy]))
+(def PolicyType (m/schema [:enum "Resource" "Identity"]))
 
-(def evaluate-one-args
+(def EvaluateOneInput
   (m/schema [:map
              [:drn :string]
              [:actionId :string]
              [:identities [:sequential {:min 1} :string]]]))
 
-(def evaluate-many-args
+(def EvaluateManyInput
   (m/schema [:map
              [:drns [:sequential {:min 1} :string]]
              [:actionId :string]
              [:identities [:sequential {:min 1} :string]]]))
 
-(defn assert*
-  ([schema obj]
-   (when-not (m/validate schema obj)
-     (throw (ex-info "Invalid type" {:anomaly/category ::invalid-type
-                                     :schema schema
-                                     :obj obj
-                                     :cause (m/explain schema obj)})))))
+(def StatementInput
+  [:map
+   [:sid {:optional true
+          :output-k :statement/sid} string?]
+   [:actionIds {:output-k :statement/actions} [:sequential string?]]
+   [:effect {:output-k (fn [x] [:statement/effect [:effect (csk/->kebab-case-keyword x)]])} string?]])
+
+(def ResourceStatementInput
+  (mu/merge StatementInput
+    [:map
+     [:identities {:output-k :statement/identities} [:sequential string?]]]))
+
+(def IdentityStatementInput
+  (mu/merge StatementInput
+    [:map
+     [:resources {:output-k :statement/resources} [:sequential string?]]]))
+
+(defn unique-sids?
+  [statements]
+  ;; we ignore `nil` sids since those do not need to be unique
+  (let [sids (keep :sid statements)]
+    (= (distinct sids) sids)))
+
+(def UniqueSids
+  [:fn {:error/fn (fn [{:keys [value]} _]
+                    (str "policy must have unique statement ids, got duplicate `sid`s: " (str/join ", " (keep :sid value))))}
+   unique-sids?])
+
+(def ResourcePolicyInput
+  [:map
+   [:drn {:output-k :policy/drn} string?]
+   [:statements {:output-k :policy/statements}
+    [:and
+     [:sequential {:min 1} ResourceStatementInput]
+     UniqueSids]]])
+
+(def IdentityPolicyInput
+  [:map
+   [:drn {:output-k :policy/drn} string?]
+   [:statements {:output-k :policy/statements}
+    [:and
+     [:sequential {:min 1} IdentityStatementInput]
+     UniqueSids]]])
+
+(def UpsertPoliciesInput
+  [:multi {:dispatch :policyType}
+   ["Resource"
+    [:map
+     [:policyType [:= "Resource"]]
+     [:policies [:sequential ResourcePolicyInput]]]]
+   ["Identity"
+    [:map
+     [:policyType [:= "Identity"]]
+     [:policies [:sequential IdentityPolicyInput]]]]])
+
+(def StatementOutput
+  [:map
+   [:statement/sid {:optional true
+                    :output-k :sid} string?]
+   [:statement/actions {:output-k :actionIds} [:sequential string?]]
+   [:statement/effect {:output-k (fn [{:keys [effect]}] [:effect (csk/->SCREAMING_SNAKE_CASE_STRING effect)])} string?]])
```

Comment:
This is part of the changes I would make to the underlying data model. We're trying to use a keyword as an enum. Typically the preferred way to do that in these Datomic-esque systems is with an ident. Perhaps a change for later.